typing/printtyp.ml

```diff
@@ -874,28 +878,59 @@ end = struct
     if non_gen then new_weak_name ty ()
     else new_name ()
 
-  let name_of_type name_generator t =
+  let name_of_type ~non_gen name_generator t =
     (* We've already been through repr at this stage, so t is our representative
        of the union-find class. *)
     let t = substitute t in
     try List.assq t !names with Not_found ->
       try TransientTypeMap.find t !weak_var_map with Not_found ->
       let name =
         match t.desc with
-          Tvar (Some name) | Tunivar (Some name) ->
-            (* Some part of the type we've already printed has assigned another
-             * unification variable to that name. We want to keep the name, so
-             * try adding a number until we find a name that's not taken. *)
+          Tvar (Some name) | Tunivar (Some name) when not non_gen ->
             let available name =
               List.for_all
                 (fun (_, name') -> name <> name')
                 !names
             in
             if available name then name
             else
-              let suffixed i = name ^ Int.to_string i in
-              let i = Misc.find_first_mono (fun i -> available (suffixed i)) in
-              suffixed i
+              (* Some part of the type we've already printed has assigned
+                 another unification variable to that name. Yet we want to
+                 keep the name we have, even though it's taken. Our approach:
+                 - If the name is a single letter, try the next 4 single
+                   letters. This will get us from 'a to 'e or from 'k to 'o.
+                   Why 4? No good reason; it just seems the right number.
+                 - Otherwise, add an increasing numeric suffix until we find an
+                   available name.
```

Comment:
It was suggested I look at the PR because of `Misc.find_first_mono`; the change proposed is fine. The use of `Misc.find_first_mono` is designed to avoid a quadratic worst-case where we draw many fresh names, and each of them goes over the list of all names before it to find the next available name. Here the change introduces a constant-time number of heuristic lookups followed by a `Misc.find_first_monod` fallback, so there is no quadratic behavior sneaking back in.